File: python/src/ehbp/session.py
"""Session recovery tokens (SPEC Section 6).

A token carries the minimal material needed to decrypt a single response
without retaining the live HPKE context: the HPKE export secret and the request
encapsulated key. It MUST be treated as sensitive key material.
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any, Union

from .derive import decrypt_framed_response, derive_response_keys
from .errors import InvalidInputError, ProtocolError
from .protocol import EXPORT_LENGTH, REQUEST_ENC_LENGTH, RESPONSE_NONCE_LENGTH

_EXPORTED_SECRET_KEY = "exportedSecret"
_REQUEST_ENC_KEY = "requestEnc"


class SessionRecoveryToken:
    __slots__ = ("_exported_secret", "_request_enc")

    def __init__(self, exported_secret: bytes, request_enc: bytes) -> None:
        if len(exported_secret) != EXPORT_LENGTH:
            raise InvalidInputError(
                f"exported secret must be {EXPORT_LENGTH} bytes, got {len(exported_secret)}"
            )
        if len(request_enc) != REQUEST_ENC_LENGTH:
            raise InvalidInputError(
                f"request enc must be {REQUEST_ENC_LENGTH} bytes, got {len(request_enc)}"
            )
        self._exported_secret = bytes(exported_secret)
        self._request_enc = bytes(request_enc)

    @property
    def exported_secret(self) -> bytes:
        return self._exported_secret

    @property
    def request_enc(self) -> bytes:
        return self._request_enc

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, SessionRecoveryToken):
            return NotImplemented
        return (
            self._exported_secret == other._exported_secret
            and self._request_enc == other._request_enc
        )

    def __hash__(self) -> int:
        return hash((self._exported_secret, self._request_enc))

    def __repr__(self) -> str:
        return "SessionRecoveryToken(exported_secret='[redacted]', request_enc='[redacted]')"

    def to_dict(self) -> dict[str, str]:
        return {
            _EXPORTED_SECRET_KEY: self._exported_secret.hex(),
            _REQUEST_ENC_KEY: self._request_enc.hex(),
        }

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

    @classmethod
    def from_dict(cls, data: Mapping[str, Any]) -> SessionRecoveryToken:
        try:
            exported_secret = bytes.fromhex(data[_EXPORTED_SECRET_KEY])
            request_enc = bytes.fromhex(data[_REQUEST_ENC_KEY])
        except (KeyError, TypeError, ValueError) as err:
            raise InvalidInputError(f"invalid session recovery token: {err}") from err
        return cls(exported_secret, request_enc)

    @classmethod
    def from_json(cls, data: Union[str, bytes]) -> SessionRecoveryToken:
        try:
            decoded = json.loads(data)
        except (TypeError, ValueError) as err:
            raise InvalidInputError(f"invalid session recovery token JSON: {err}") from err
        if not isinstance(decoded, Mapping):
            raise InvalidInputError("invalid session recovery token: expected JSON object")
        return cls.from_dict(decoded)

    def decrypt_response_body(self, response_nonce: bytes, body: bytes) -> bytes:
        if len(response_nonce) != RESPONSE_NONCE_LENGTH:
            raise ProtocolError(
                f"response nonce must be {RESPONSE_NONCE_LENGTH} bytes, got {len(response_nonce)}"
            )
        key_material = derive_response_keys(
            self._exported_secret, self._request_enc, response_nonce
        )
        return decrypt_framed_response(key_material, body)

File: rust/src/session.rs
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::fmt;
use zeroize::{Zeroize, Zeroizing};

use crate::{
    derive::{decrypt_framed_response, derive_response_keys},
    protocol::{EXPORT_LENGTH, REQUEST_ENC_LENGTH, RESPONSE_NONCE_LENGTH},
    Error, Result,
};

#[derive(Clone, Eq, PartialEq)]
pub struct SessionRecoveryToken {
    pub exported_secret: Vec<u8>,
    pub request_enc: Vec<u8>,
}

struct SensitiveString(Zeroizing<String>);

impl<'de> Deserialize<'de> for SensitiveString {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(|value| Self(Zeroizing::new(value)))
    }
}

impl fmt::Debug for SessionRecoveryToken {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("SessionRecoveryToken")
            .field("exported_secret", &"[redacted]")
            .field("request_enc", &"[redacted]")
            .finish()
    }
}

impl Drop for SessionRecoveryToken {
    fn drop(&mut self) {
        self.exported_secret.zeroize();
        self.request_enc.zeroize();
    }
}

impl SessionRecoveryToken {
    pub fn new(exported_secret: Vec<u8>, request_enc: Vec<u8>) -> Result<Self> {
        if exported_secret.len() != EXPORT_LENGTH {
            return Err(Error::InvalidInput(format!(
                "exported secret must be {EXPORT_LENGTH} bytes, got {}",
                exported_secret.len()
            )));
        }
        if request_enc.len() != REQUEST_ENC_LENGTH {
            return Err(Error::InvalidInput(format!(
                "request enc must be {REQUEST_ENC_LENGTH} bytes, got {}",
                request_enc.len()
            )));
        }
        Ok(Self {
            exported_secret,
            request_enc,
        })
    }

    pub fn decrypt_response_body(&self, response_nonce: &[u8], body: &[u8]) -> Result<Vec<u8>> {
        if response_nonce.len() != RESPONSE_NONCE_LENGTH {
            return Err(Error::Protocol(format!(
                "response nonce must be {RESPONSE_NONCE_LENGTH} bytes, got {}",
                response_nonce.len()
            )));
        }
        let key_material =
            derive_response_keys(&self.exported_secret, &self.request_enc, response_nonce)?;
        decrypt_framed_response(&key_material, body)
    }
}

impl Serialize for SessionRecoveryToken {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Token<'a> {
            #[serde(rename = "exportedSecret")]
            exported_secret: &'a str,
            #[serde(rename = "requestEnc")]
            request_enc: &'a str,
        }

        let exported_secret = Zeroizing::new(hex::encode(&self.exported_secret));
        let request_enc = Zeroizing::new(hex::encode(&self.request_enc));
        Token {
            exported_secret: &exported_secret,
            request_enc: &request_enc,
        }
        .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for SessionRecoveryToken {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct Token {
            #[serde(rename = "exportedSecret")]
            exported_secret: SensitiveString,
            #[serde(rename = "requestEnc")]
            request_enc: SensitiveString,
        }

        let token = Token::deserialize(deserializer)?;
        let exported_secret = Zeroizing::new(
            hex::decode(token.exported_secret.0.as_bytes()).map_err(serde::de::Error::custom)?,
        );
        let request_enc = Zeroizing::new(
            hex::decode(token.request_enc.0.as_bytes()).map_err(serde::de::Error::custom)?,
        );
        if exported_secret.len() != EXPORT_LENGTH {
            return Err(serde::de::Error::custom(format!(
                "exported secret must be {EXPORT_LENGTH} bytes, got {}",
                exported_secret.len()
            )));
        }
        if request_enc.len() != REQUEST_ENC_LENGTH {
            return Err(serde::de::Error::custom(format!(
                "request enc must be {REQUEST_ENC_LENGTH} bytes, got {}",
                request_enc.len()
            )));
        }
        Ok(Self {
            exported_secret: exported_secret.to_vec(),
            request_enc: request_enc.to_vec(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn debug_redacts_session_recovery_token_material() {
        let token = SessionRecoveryToken::new(vec![1; 32], vec![2; 32]).unwrap();
        let debug = format!("{token:?}");

        assert!(debug.contains("[redacted]"));
        assert!(!debug.contains("1, 1"));
        assert!(!debug.contains("2, 2"));
    }
}
